engine/runtime/action-source/src/file/obj.rs

```diff
@@ -922,4 +1060,59 @@ f 4 5 6
         assert_eq!(obj_data.faces[1].object, Some("Pyramid".to_string()));
         assert_eq!(obj_data.faces[1].group, Some("pyramid_group".to_string()));
     }
+
+    #[test]
+    fn test_safe_f64_to_number() {
+        assert_eq!(
+            safe_f64_to_number(1.5),
+            serde_json::Number::from_f64(1.5).unwrap()
+        );
+        assert_eq!(
+            safe_f64_to_number(0.0),
+            serde_json::Number::from_f64(0.0).unwrap()
+        );
+        assert_eq!(
+            safe_f64_to_number(-1.5),
+            serde_json::Number::from_f64(-1.5).unwrap()
+        );
+
+        assert_eq!(safe_f64_to_number(f64::NAN), serde_json::Number::from(0));
+        assert_eq!(
+            safe_f64_to_number(f64::INFINITY),
+            serde_json::Number::from(0)
+        );
+        assert_eq!(
+            safe_f64_to_number(f64::NEG_INFINITY),
+            serde_json::Number::from(0)
+        );
+    }
+
+    #[test]
+    fn test_material_properties_with_edge_values() {
+        let mut materials = HashMap::new();
+        let mat = Material {
+            name: "test_mat".to_string(),
+            diffuse: Some([1.0, f32::NAN, 0.5]),
+            shininess: Some(f32::INFINITY),
+            ..Default::default()
+        };
+        materials.insert("test_mat".to_string(), mat);
+
+        let used_materials = vec!["test_mat".to_string()];
+        let (_, material_properties) = extract_material_properties(&materials, &used_materials);
+
+        if let AttributeValue::Map(props) = material_properties {
```

Comment:
I think replacing the `if let` patterns with `let...else` would be better, because the current approach can silently pass if the expected structure isn't found.

```rust
let AttributeValue::Map(props) = material_properties else {
    panic!("Expected material_properties to be a Map");
};

[...]
```

This way the test will explicitly fail with a clear message instead of potentially passing without testing anything. What do you think?